### DISCUSSION TITLE
Error in post-command-hook (vertico--exhibit): (wrong-type-argument stringp t)

### DISCUSSION THREAD

**Discussion by arouene** [minad/vertico#503] [Q&A] [answered]
Similar to issue #500, but reinstalling vertico does not solve the issue (I reinstalled all the packages).

![image](https://github.com/user-attachments/assets/fa3c6ed2-e3cb-470d-abb7-f7081671351f)

I'm actually trying to debug it. For now, I know that it comes from 

```
(overlay-put vertico--count-ov 'before-string
             (if vertico-count-format (vertico--format-count) "")))
```

in **vertico--display count**

**Comment by minad** [comment]
Do you have a minimal recipe starting from emacs -Q? See also https://github.com/minad/vertico?tab=readme-ov-file#debugging-vertico.

**Reply by arouene** [reply]
Yes, I finally got one :

```
(require 'package)

(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(add-to-list 'package-archives '("gnu" . "https://elpa.gnu.org/packages/") t)
(add-to-list 'package-archives '("nongnu" . "https://elpa.nongnu.org/nongnu/") t)

(package-initialize)

(use-package vertico
  :ensure t
  :bind (:map vertico-map
              ("<backspace>" . #'vertico-directory-delete-char))
  :init
  (vertico-mode))

(use-package marginalia
  :ensure t
  :init
  (marginalia-mode))

(use-package kanagawa-theme
  :ensure t)
```

And then C-h v
Start typing "kanagawa"
At "ka" it display an error and then vertico stops working.

**Reply by arouene** [reply]
Somehow this is the interaction between 'kanagawa-theme-true-color-p and marginalia

```
(marginalia-annotate-variable 'kanagawa-theme-true-color-p)
;; produce
Debugger entered--Lisp error: (wrong-type-argument stringp t)
  string-search("\n" t)
  (and t (string-search "\n" str))
  ...
```

**Reply by arouene** [reply, marked as answer]
In this var, the doctring and initial value is inverted.

```
(defvar kanagawa-theme-true-color-p
  "When non-nil use true color values in color palette."
  (or (display-graphic-p) (= (tty-display-color-cells) 16777216)))
```

So marginalia get this as a type "string", but later it returns a bool.

The "fault" is obviously from kanagawa-theme. But it could be nice if vertico/marginalia handles the problem gracefully, as one variable in the completion modeline can make vertico stop working, even if this is not the variable we are searching for.

**Reply by arouene** [reply]
Note aside, I opened an issue on kanagawa-theme : https://github.com/Fabiokleis/emacs-kanagawa-theme/issues/5

**Reply by minad** [reply]
> The "fault" is obviously from kanagawa-theme. But it could be nice if vertico/marginalia handles the problem gracefully, as one variable in the completion modeline can make vertico stop working, even if this is not the variable we are searching for.

There is no point where we can stop if we start to handle arbitrary type errors like this one gracefully. Imo it makes sense to fail hard here such that bugs get fixed quickly at the root. If we stop failing on type errors like this one, we subvert the type system and may hide even worse problems later on.

Btw, one can easily check everything via `embark-collect`. For example invoke `describe-symbol` and then invoke `embark-collect` such that Marginalia generates annotations for all symbols. If something fails, there is a bug somewhere. I used this feature quite often during the development of Marginalia.

**Reply by arouene** [reply]
> If we stop failing on type errors like this one, we subvert the type system and may hide even worse problems later on.

Yes it make sense :+1: 

> Btw, one can easily check everything via embark-collect. For example invoke describe-symbol and then invoke embark-collect such that Marginalia generates annotations for all symbols. If something fails, there is a bug somewhere. I used this feature quite often during the development of Marginalia.

Nice to know, thank you ! (embark is really a piece of art)
